### PR TITLE
Automatically propagate inherited environment to child blueprint views

### PR DESCRIPTION
**Pull request by kyleve**
This PR introduces automatic propagation of `Environment` through to arbitrarily nested `BlueprintViews`. This is useful, eg in `Market`, where in many places we nest additional `BlueprintViews` within stateful controls such as buttons – previously `Environment` values either would not flow down to these nested views, or we would have to manually flow the `Environment` down to these views.

We accomplish this propagation through updates to `NativeViewNode` and `NativeViewController`. When we create a `NativeViewNode`, we now record its `Environment`. Then, when the `NativeViewController` goes to update the `node`'s associated `UIView`, we assign the `UIView`'s `nativeViewNodeBlueprintEnvironment` (added via associated objects). This means that the `UIView` in the view hierarchy now has the correct `Environment`, which corresponds to its `Element`'s environment at measurement / layout time. 

To fetch our inherited `Environment`, when it comes time for a `BlueprintView` to layout its `Element` hierarchy; it asks for its own `inheritedBlueprintEnvironment` – this walks up the superview hierarchy until we find a `nativeViewNodeBlueprintEnvironment`, signaling that we are inside a parent `BlueprintView`; and uses this `Environment` to seed the `makeEnvironment()` function.

### CODE REVIEW

File: BlueprintUI/Sources/BlueprintView/BlueprintView.swift [hunk 1]
```diff
@@ -37,13 +37,23 @@ public final class BlueprintView: UIView {
...
             invalidateIntrinsicContentSize()
         }
     }
+    
+    /// A base environment inherited when this ``BlueprintView`` is within the view hierarchy of another ``BlueprintView``.
+    /// Set during the view layout portion of a layout cycle.
+    /// Some keys may be overridden based on the traits of the view itself, or based on the explicit environment set on `environment`.
+    public internal(set) var inheritedEnvironment : Environment? {
+        didSet {
+            setNeedsViewHierarchyUpdate()
```
Comment (kyleve) [inline BlueprintUI/Sources/BlueprintView/BlueprintView.swift:53]: Actually; I don't think I need this; since the layout pass itself should do this?

File: BlueprintUI/Sources/BlueprintView/BlueprintView.swift [hunk 2]
```diff
@@ -403,7 +434,7 @@ extension BlueprintView {
                             contentView.insertSubview(controller.view, at: index)
                         }
 
-                        controller.update(node: child, appearanceTransitionsEnabled: true)
+                        controller.update(node: child, context: context)
```
Comment (kyleve) [inline BlueprintUI/Sources/BlueprintView/BlueprintView.swift:437]: Passing `true` all the time here was wrong; but also a no-op kind of wrong; since we'd never have any children to update if we hadn't done at least one update path.

File: BlueprintUI/Sources/BlueprintView/BlueprintView.swift [hunk 3]
```diff
@@ -441,3 +474,21 @@ extension BlueprintView {
         }
     }
 }
+
+extension BlueprintView.NativeViewController {
+    
+    /// A context value passed to the ``BlueprintView/NativeViewController`` instance as the view tree is updated by blueprint.
+    /// Contains information relevant to the correct construction and management of the view tree.
+    struct UpdateContext {
```
Comment (kyleve) [inline BlueprintUI/Sources/BlueprintView/BlueprintView.swift:482]: This isn't a needed change; but I did it since it shortens up the callsites and makes it easier to add additional information here without appending tons of parameters (like animation state, which I was planning on looking at reviving soon).

File: SampleApp/Podfile.lock [hunk 4]
```diff
@@ -2,7 +2,7 @@ PODS:
   - BlueprintUI (0.27.0)
   - BlueprintUI/Tests (0.27.0)
   - BlueprintUICommonControls (0.27.0):
-    - BlueprintUI
+    - BlueprintUI (= 0.27.0)
```
Comment (kyleve) [inline SampleApp/Podfile.lock:5]: Out of date on `main`.

File: BlueprintUI/Sources/Environment/Environment.swift [hunk 5]
```diff
@@ -69,8 +69,50 @@ public struct Environment {
...
+
+
+extension UIView {
+    
+    /// The ``Environment`` for the ``Element`` that this view represents in a Blueprint element tree,
+    /// or if the view is not explicitly managed by Blueprint, the ``Environment`` of
+    /// the nearest superview that is managed by Blueprint.
+    ///
+    /// If no views in the superview hierarchy are managed by Blueprint, this property returns nil.
+    var inheritedBlueprintEnvironment : Environment? {
```
Comment (kyleve) [inline BlueprintUI/Sources/Environment/Environment.swift:95]: I kept this internal for now; but would this be helpful at all, eg in the Market view layer; would it simplify anything?

File: BlueprintUI/Tests/EnvironmentTests.swift [hunk 6]
```diff
@@ -132,13 +137,26 @@ class EnvironmentTests: XCTestCase {
...
+    func test_merged() {
+        var lhs = Environment.empty
+        lhs[TestKey.self] = .wrong
+        
+        var rhs = Environment.empty
+        rhs[TestKey.self] = .right
+        
+        let output = lhs.merged(prioritizing: rhs)
+        
+        XCTAssertEqual(output[TestKey.self], .right)
```
Comment (n8chur) [inline BlueprintUI/Tests/EnvironmentTests.swift:158]: Would it also be worth testing that setting values for properties exclusively on each side are merged correctly? E.g. only set `KeyA` on `lhs` and `KeyB` on `rhs` and validate that both values exist on the merged object.

File: BlueprintUI/Tests/EnvironmentTests.swift [hunk 7]
```diff
@@ -163,6 +181,47 @@ class EnvironmentTests: XCTestCase {
...
+        
+        XCTAssertEqual(first.inheritedBlueprintEnvironment?[TestingKey1.self], 1)
+        XCTAssertEqual(second.inheritedBlueprintEnvironment?[TestingKey1.self], 1)
+        XCTAssertEqual(third.inheritedBlueprintEnvironment?[TestingKey1.self], 1)
+        
+        first.nativeViewNodeBlueprintEnvironment = nil
+        
+        XCTAssertEqual(first.inheritedBlueprintEnvironment?[TestingKey1.self], nil)
+        XCTAssertEqual(second.inheritedBlueprintEnvironment?[TestingKey1.self], nil)
+        XCTAssertEqual(third.inheritedBlueprintEnvironment?[TestingKey1.self], nil)
```
Comment (n8chur) [inline BlueprintUI/Tests/EnvironmentTests.swift:206]: Should we also test setting a non-nil value on a non-root view and ensure that superviews are not resolving to a value?
```suggestion
        XCTAssertEqual(third.inheritedBlueprintEnvironment?[TestingKey1.self], nil)
        
        second.nativeViewNodeBlueprintEnvironment = 2

        XCTAssertEqual(first.inheritedBlueprintEnvironment?[TestingKey1.self], nil)
        XCTAssertEqual(second.inheritedBlueprintEnvironment?[TestingKey1.self], 2)
        XCTAssertEqual(third.inheritedBlueprintEnvironment?[TestingKey1.self], 2)
        
```

**Review by n8chur — APPROVED** [review]

File: BlueprintUI/Sources/BlueprintView/BlueprintView.swift [hunk 8]
```diff
@@ -268,7 +300,19 @@ public final class BlueprintView: UIView {
     }
     
     private func makeEnvironment() -> Environment {
-        var environment = self.environment
+                
+        let inherited : Environment = {
+            if
+                let inherited = self.inheritedBlueprintEnvironment,
+                self.automaticallyInheritsEnvironmentFromContainingBlueprintViews
```
[the remark this answers is not in the document]
Comment (kyleve) [reply BlueprintUI/Sources/BlueprintView/BlueprintView.swift:307]: Good call

File: BlueprintUI/Sources/BlueprintView/BlueprintView.swift [hunk 9]
```diff
@@ -323,20 +366,28 @@ extension BlueprintView {
             self.viewDescription = node.viewDescription
             self.layoutAttributes = node.layoutAttributes
             self.children = []
+            
             self.view = node.viewDescription.build()
+            self.view.nativeViewNodeBlueprintEnvironment = node.environment
+        }
+        
+        deinit {
+            self.view.nativeViewNodeBlueprintEnvironment = nil
```
[the remark this answers is not in the document]
Comment (kyleve) [reply BlueprintUI/Sources/BlueprintView/BlueprintView.swift:375]: This is a belt-and-suspenders thing – in _theory_ you should always be returning a new view from your view description; but you don't _have_ to – it could be a static / instance owned elsewhere. So I am clearing the value out when we stop controlling the view just in case.

File: BlueprintUI/Sources/BlueprintView/BlueprintView.swift [hunk 10]
```diff
@@ -37,13 +37,32 @@ public final class BlueprintView: UIView {
     private let rootController: NativeViewController
 
     /// A base environment to render elements with.
-    /// Some keys may be overriden based on the traits of the view itself.
+    /// Some keys may be overridden based on the traits of the view itself.
```
[the remark this answers is not in the document]
Comment (kyleve) [reply BlueprintUI/Sources/BlueprintView/BlueprintView.swift:40]: I updated the docs here:

```
    /// A base environment used when laying out and rendering the element tree.
    ///
    /// Some keys will be overridden with the traits from the view itself. Eg, `windowSize`, `safeAreaInsets`, etc.
    ///
    /// If this blueprint view is within another blueprint view, the environment of the parent view
    /// will be inherited by this view if `automaticallyInheritsEnvironmentFromContainingBlueprintViews`
    /// is enabled. In the case of matching keys in both the inherited environment and the provided
    /// environment, the values from this environment will take priority over the inherited environment.
```

File: BlueprintUI/Tests/EnvironmentTests.swift (hunk 6 above)
Comment (kyleve) [reply BlueprintUI/Tests/EnvironmentTests.swift:158]: Sure

File: BlueprintUI/Tests/EnvironmentTests.swift (hunk 7 above)
Comment (kyleve) [reply BlueprintUI/Tests/EnvironmentTests.swift:206]: I'm not quite sure I follow what you mean here – can you expand a bit?

File: BlueprintUI/Tests/EnvironmentTests.swift (hunk 6 above)
Comment (kyleve) [reply BlueprintUI/Tests/EnvironmentTests.swift]: Done

File: BlueprintUI/Tests/EnvironmentTests.swift (hunk 7 above)
Comment (kyleve) [reply BlueprintUI/Tests/EnvironmentTests.swift:214]: (If I'm reading the comment correctly; the concern I think is a superview finding a value from a subview? As implemented; that can't happen, since we only traverse up.)

Comment (n8chur) [reply BlueprintUI/Tests/EnvironmentTests.swift:214]: Yeah, I don't see how the test I'm suggesting would fail—no big deal if you want to omit it. It's really just asserting that you're not using a reference to the same environment in each layer of the tree.
